src/components/legend.ts

```diff
@@ -237,88 +514,107 @@ namespace Plottable.Components {
-        let lowY = i * layout.textHeight + legendPadding;
-        let highY = (i + 1) * layout.textHeight + legendPadding;
-        let symbolY = (lowY + highY) / 2;
-        let lowX = legendPadding;
-        let highX = legendPadding;
-        d3.select(this).selectAll("g." + Legend.LEGEND_ENTRY_CLASS).each(function(value: string) {
-          highX += layout.entryLengths.get(value);
-          let symbolX = lowX + layout.textHeight / 2;
-          if (highX >= p.x && lowX <= p.x &&
-              highY >= p.y && lowY <= p.y) {
-            let entrySelection = d3.select(this);
-            let datum = entrySelection.datum();
-            entities.push({
-              datum: datum,
-              position: { x: symbolX, y: symbolY },
-              selection: entrySelection,
-              component: legend,
-            });
+      function within(p: Point, bounds: Bounds) {
+        return bounds.topLeft.x <= p.x
+          && bounds.bottomRight.x >= p.x
+          && bounds.topLeft.y <= p.y
+          && bounds.bottomRight.y >= p.y;
+      };
+
+      const table = this._buildLegendTable(this.width(), this.height());
+      return table.reduce((entity: Entity<Legend>[], row: LegendRow, rowIndex: number) => {
+        if (entity.length !== 0) {
+          // we've already found the nearest entity; just return it.
+          return entity;
+        }
+
+        const rowBounds = table.getBounds(row);
+        const withinRow = within(p, rowBounds);
+
+        if (!withinRow) {
+          // the nearest entity isn't within this row, continue;
+          return entity;
+        }
+
+        return row.reduce((entity: Entity<Legend>[], column: LegendColumn<string>, columnIndex: number) => {
+          const columnBounds = table.getBounds(row, column);
+          const withinColumn = within(p, columnBounds);
+
+          if (withinColumn) {
+            const rowElement = this.content().selectAll(`.${Legend.LEGEND_ROW_CLASS}`)[0][rowIndex];
+            // HACKHACK The 2.x API chooses the symbol element as the "selection" to return, regardless of what
+            // was actually selected
+            const entryElement = d3.select(rowElement)
+              .selectAll(`.${Legend.LEGEND_ENTRY_CLASS}`)[0][Math.floor(columnIndex / 2)];
+             const symbolElement = d3.select(entryElement).select(`.${Legend.LEGEND_SYMBOL_CLASS}`);
+
+            // HACKHACK The 2.x API returns the center {x, y} of the symbol as the position.
+            const rowTranslate = d3.transform(d3.select(rowElement).attr("transform")).translate;
+            const symbolTranslate = d3.transform(symbolElement.attr("transform")).translate;
+
+            return [{
+              datum: column.getData(),
+              position: { x: rowTranslate[0] + symbolTranslate[0], y: rowTranslate[1] + symbolTranslate[1] },
+              selection: d3.select(entryElement),
+              component: this,
+            }];
```

Comment:
and here instead of pushing to `entity`, we just return a statically declared single item array, which seems weird that we've tried to pass around the initial empty array through all the reduce calls.